gstreamer/src/query.rs

```diff
@@ -15,114 +15,88 @@ use std::ptr;
-    }
-
-    pub fn new_latency() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_latency()) }
-    }
-
-    pub fn new_seeking(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_seeking(fmt.to_glib())) }
-    }
+pub struct GstRc2<V: View>(V);
 
-    pub fn new_segment(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_segment(fmt.to_glib())) }
+impl<V: View> GstRc2<V> {
+    pub fn get_view<'a>(&'a self) -> &'a V {
+        &self.0
     }
 
-    pub fn new_convert<V: Into<GenericFormattedValue>>(value: V, dest_fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        let value = value.into();
-        unsafe {
-            from_glib_full(ffi::gst_query_new_convert(
-                value.get_format().to_glib(),
-                value.get_value(),
-                dest_fmt.to_glib(),
-            ))
+    pub fn try_view_mut<'a>(&'a mut self) -> Option<&'a mut V> {
+        if self.0.is_writable() {
+            Some(&mut self.0)
+        } else {
+            None
         }
     }
 
-    pub fn new_formats() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_formats()) }
-    }
-
-    pub fn new_buffering(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_buffering(fmt.to_glib())) }
+    pub fn try_generic_mut<'a>(&'a mut self) -> Option<&'a mut V::Generic> {
+        if self.0.is_writable() {
+            Some(self.0.as_mut())
+        } else {
+            None
+        }
     }
 
-    pub fn new_custom(structure: ::Structure) -> Self {
-        assert_initialized_main_thread!();
-        unsafe {
-            from_glib_full(ffi::gst_query_new_custom(
-                ffi::GST_QUERY_CUSTOM,
-                structure.into_ptr(),
-            ))
+    pub fn try_into_owned(self) -> Option<V::Generic> {
+        if self.0.is_writable() {
+            Some(self.0.into_owned())
+        } else {
+            None
         }
     }
+}
 
-    pub fn new_uri() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_uri()) }
-    }
+impl<V: View> Deref for GstRc2<V> {
+    type Target = V;
 
-    pub fn new_scheduling() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_scheduling()) }
+    fn deref(&self) -> &Self::Target {
+        &self.0
     }
+}
 
-    pub fn new_accept_caps(caps: &::Caps) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_accept_caps(caps.as_mut_ptr())) }
-    }
+// *** Architecture Preview ***
+// Query specific definitions start here
+pub struct QueryRef(*mut ffi::GstQuery);
```

Comment:
This is always an owned pointer, or how would you distinguish between owned and borrowed here? Borrowing without owning is needed for the pad functions for example

Also I think this is unsafe. Whenever you provide a &mut QueryRef now, it would be possible to replace it. But that would mean that the pointer would be replaced, potentially causing interesting memory unsafety issues with regard to lifetimes of the underlying object.
That's what the QueryRef (and the other Ref types) are the actual C types, so that getting a mutable reference would replace the underlying struct instead of the pointer.